src/include/cursor.i

```diff
@@ -201,16 +201,16 @@ __cursor_reset(WT_CURSOR_BTREE *cbt)
     if (F_ISSET(cbt, WT_CBT_ACTIVE)) {
         if (!F_ISSET(cbt, WT_CBT_NO_TRACKING))
             __cursor_leave(session);
+
+        /*
+         * When the count of active cursors in the session goes to zero, there are no active
+         * cursors, and we can release any snapshot we're holding.
+         */
+        if (session->ncursors == 0 && !F_ISSET(cbt, WT_CBT_NO_TXN))
+            __wt_txn_read_last(session);
         F_CLR(cbt, WT_CBT_ACTIVE);
     }
 
-    /*
-     * When the count of active cursors in the session goes to zero, there are no active cursors,
-     * and we can release any snapshot we're holding for read committed isolation.
-     */
-    if (session->ncursors == 0 && !F_ISSET(cbt, WT_CBT_NO_TXN))
```

Comment:
It's taken me a very long time to understand what you are trying to achieve with this change. Is the problem that there is a session coming into __cursor_reset without an active btree cursor that is relying on keeping a snapshot alive even though it doesn't have any cursors open and isn't running at snapshot isolation?
That seems like an unreasonable expectation on __cursor_reset.

Reply:
The transaction is supposed to run at snapshot isolation. But when it comes to this code, it hasn't got a snapshot yet.

Reply:
```
/*
 * __wt_txn_begin --
 *     Begin a transaction.
 */
static inline int
__wt_txn_begin(WT_SESSION_IMPL *session, const char *cfg[])
{
    WT_TXN *txn;

    txn = session->txn;
    txn->isolation = session->isolation;
    txn->txn_logsync = S2C(session)->txn_logsync;

    WT_ASSERT(session, !F_ISSET(txn, WT_TXN_RUNNING));

    WT_RET(__wt_txn_config(session, cfg));

    /* Allocate a snapshot if required. */
    if (txn->isolation == WT_ISO_SNAPSHOT) {
        if (session->ncursors > 0)
            WT_RET(__wt_session_copy_values(session));

        /*
         * Stall here if the cache is completely full. We have allocated a transaction ID which
         * makes it possible for eviction to decide we're contributing to the problem and return
         * WT_ROLLBACK. The WT_SESSION.begin_transaction API can't return rollback, continue on.
         */
        WT_RET_ERROR_OK(__wt_cache_eviction_check(session, false, true, NULL), WT_ROLLBACK);

        __wt_txn_get_snapshot(session);
    }

    F_SET(txn, WT_TXN_RUNNING);
    if (F_ISSET(S2C(session), WT_CONN_READONLY))
        F_SET(txn, WT_TXN_READONLY);

    return (0);
}
```

The problem happens in the eviction before getting the snapshot:

```
        /*
         * Stall here if the cache is completely full. We have allocated a transaction ID which
         * makes it possible for eviction to decide we're contributing to the problem and return
         * WT_ROLLBACK. The WT_SESSION.begin_transaction API can't return rollback, continue on.
         */
        WT_RET_ERROR_OK(__wt_cache_eviction_check(session, false, true, NULL), WT_ROLLBACK);

```

So my first attempt to fix the problem is to mark the transaction as WT_TXN_RUNNING before doing this eviction.